### ISSUE TITLE
add Python 3 support

### ISSUE THREAD

**Issue by tomkralidis** [geopython/pycsw#328]
pycsw currently supports 2.6 and 2.7 (as per https://github.com/geopython/pycsw/blob/master/.travis.yml#L3).

Add support for Python 3 and include as part of travis-ci testing.

**Comment by kalxas** [comment]
This should be on the roadmap for pycsw 3.x

**Comment by kalxas** [comment]
https://github.com/geopython/pycsw/pull/376

Since we have an active PR in this issue, I think we should wait for this feature to be included in pycsw 2.0.

Also, since it is a major feature we should provide a RFC for it.

**Comment by tomkralidis** [comment]
:+1:

**Comment by ricardogsilva** [comment]
Hi guys

Just to clarify, is this is to be postponed until after pycsw2.0?

If so, I think I can pick up the new py3k branch and fix the remaining issues (at a smaller pace). I'd also like to take a softer approach than the one in use, mostly by:

* being more selective with the `import` statements (the current state of the py3k branch has lots of 
`from __future__ import (absolute_import, division, print_function)` in places where these are not needed, for example

* wherever possible trying to adapt the code to be compatible at the same time with Python 2 and 3. From what I've read it seems that [python-future](http://python-future.org/index.html) is a nicer compatibility layer than [six](http://pythonhosted.org/six/) and I'd like to try it out

This raises some questions though:

* To what minimal Python 2 version are we committed? According to the official porting guide, it is much easier to port the code if we decide to [only worry about supporting Python 2.7](https://docs.python.org/3/howto/pyporting.html#drop-support-for-python-2-6-and-older). In my opinion we should follow the official lead (of the Python Software Foundation) and support only Python2.7+.

* Do we wish to support Python3 (< 3.3)? [It seems that it is easier](http://python-future.org/faq.html#which-versions-of-python-does-python-future-support) and more cost effective to go with Python 3 (>= 3.3). 

So, to be clear it seems to me that the best strategy would be to support Python2 (>=2.7) and Python3 (>=3.3). I guess this is probably an issue for the PSC, so maybe a more formal discussion might be in order. What do you think?

**Comment by ocefpaf** [comment]
> being more selective with the import statements (the current state of the py3k branch has lots of 
from `__future__ import (absolute_import, division, print_function)` in places where these are not needed, for example

These do not add any extra "cost" and avoid future mistakes.

> wherever possible trying to adapt the code to be compatible at the same time with Python 2 and 3. From what I've read it seems that python-future is a nicer compatibility layer than six and I'd like to try it out

I used both in the past and I really like the six approach.  However, I have no real reason besides "I like six more."  I guess `python-future` is OK if the core devs want to pursue that.

> To what minimal Python 2 version are we committed? According to the official porting guide, it is much easier to port the code if we decide to only worry about supporting Python 2.7. In my opinion we should follow the official lead (of the Python Software Foundation) and support only Python2.7+.

That is a decision for the core developers, but as I user I would like to see pycsw for 2.6 out there for some time.  (There are some old Centos based servers out there with python 2.6 that need an updated pycsw.)

> Do we wish to support Python3 (< 3.3)? It seems that it is easier and more cost effective to go with Python 3 (>= 3.3).

Huge :+1: on this.  Python 3.2 and 3.3, IMHO, where a mistake.  Python 3 starts at 3.4 :wink:

**Comment by ocefpaf** [comment]
> From what I've read it seems that python-future is a nicer compatibility layer than six and I'd like to try it out.

I just remembered that some of `pycsw` dependencies, like `OWSLib`, already use `six`.  That means, using `python-future` adding an extra dependency, while using `six` we are keep the "status quo."

**Comment by kalxas** [comment]
> To what minimal Python 2 version are we committed? According to the official porting guide, it is much easier to port the code if we decide to only worry about supporting Python 2.7. In my opinion we should follow the official lead (of the Python Software Foundation) and support only Python2.7+.

> That is a decision for the core developers, but as I user I would like to see pycsw for 2.6 out there for some time.  (There are some old Centos based servers out there with python 2.6 that need an updated pycsw.)

Actually some of our most important deployments are using Python 2.6 under CentOS 6, so we should not kick Python 2.6 out yet...

**Comment by kalxas** [comment]
link https://github.com/geopython/pycsw/pull/392

**Comment by kalxas** [comment]
We are planning to merge py3k branch to master, without enabling python 3 tests yet https://github.com/geopython/pycsw/pull/417

This way, we can finalize python 3 support in a second iteration.

**Comment by tomkralidis** [comment]
Update: the current PR is #419.

All Python 2.6/2.7 tests pass.  There are **2** Python 3.4 remaining errors of type https://travis-ci.org/geopython/pycsw/jobs/123350424#L1095.

More details:

- request that causes the exception in Python 3.4 but not Python 2.7: https://github.com/geopython/pycsw/blob/master/tests/suites/cite/post/19d2a6ed-be28-4866-ae15-e3bb634486cb.xml
- code trace which throws the error
 - https://github.com/geopython/pycsw/blob/master/pycsw/ogc/csw/csw2.py#L1528
- https://github.com/geopython/pycsw/blob/master/pycsw/server.py#L228

I'm guessing this is rooted in the input handling of the XML bytes/strings dance which fails in Python 3.4.

**Comment by kalxas** [comment]
Excellent progress!!!
Thanks @tomkralidis

**Comment by kalxas** [comment]
I suggest we squash #419 and merge.

**Comment by tomkralidis** [comment]
@kalxas good plan.  There are some minor issues w/ CITE.  I'll work through those once the 2 errors are clear.

**Comment by tomkralidis** [comment]
cc @QuLogic

Digging deeper, Python 2 is parsing `<type 'str'>` whereas Python 3 is parsing `<class 'bytes'>.`, which is likely the issue.

**Comment by kalxas** [comment]
if CITE tests are failing we should wait...

**Comment by QuLogic** [comment]
The XML document is invalid; that's the sort of problem that Python 2's lax handling of encoding lets slip by. The encoding is specified as UTF-8, and the literal is `b"Fusc\xe9 vitae ligul\xe4"`. Now, `0xe9` and `0xe4` are the Unicode _code points_ for 'é' and 'ä', but they are _not_ valid UTF-8 encoded bytes. 'éä' encoded in UTF-8 is `b'\xc3\xa9\xc3\xa4'`.

**Comment by QuLogic** [comment]
However, the document in git is correct, so the question is, how did `self.environ['wsgi.input']` get set?

**Comment by tomkralidis** [comment]
AFAICT it comes from the WSGI env.  Perhaps we detect the type, and if bytes then we decode()?

**Comment by tomkralidis** [comment]
Does six provide anything to help bridge this?

**Comment by QuLogic** [comment]
My reading of the [WSGI spec](https://www.python.org/dev/peps/pep-3333/#unicode-issues) is that things are a bit odd. The bytestrings are _always_ ISO-8859-1 (which makes sense since those are the correct byte values for the literal.) So if you get a string from WSGI, you should _always_ decode as ISO-8859-1.

Also, I see that `Csw.environ` can sometimes be a copy of `os.environ`; I think that might be a bit fragile, encoding-wise, since it's likely not to be ISO-8859-1. The WSGI module does provide [a support function](https://docs.python.org/3/library/wsgiref.html#wsgiref.handlers.read_environ) for correctly reading from the environment.

**Comment by tomkralidis** [comment]
Thanks @QuLogic for the analysis.  @ricardogsilva any further comments / ideas here, given the recent refactor of `pycsw.server.dispatch()` and `pycsw.server.dispatch_wsgi()`?

**Comment by ricardogsilva** [comment]
@tomkralidis , @QuLogic comments seem spot on the problem.

From my understanding, and without having tested yet, it seems that @QuLogic suggestion of using `wsgiref.handlers.read_environ()` does not really apply here. 

In `pycsw/csw.py` (the CGI wrapper) we are using `wsgiref.handlers.CGIHandler`. This class is supposed to already be using `wsgiref.handlers.read_environ()` internally, as implied in the [docs](https://docs.python.org/3/library/wsgiref.html#wsgiref.handlers.read_environ). So it seems that we should be getting strings that have already been appropriately digested by wsgi. 

I've not been following this problem up until now, but from glancing at the results shown by @tomkralidis it seems that we should be explicit with the encoding used to:


1. Decode the `wsgi.input` data in https://github.com/geopython/pycsw/blob/master/pycsw/server.py#L229

 I guess we need to assume a default encoding when first getting the request.input from wsgi's environment. This default encoding should come from:

  1. The HTTP Content-Type header, if available
  2. Whatever wsgi may use by default (I have to catch up on PEP 3333)

2. Later, when we try to convert the string to an `etree.Element` in https://github.com/geopython/pycsw/blob/master/pycsw/ogc/csw/csw2.py#L1530 

In any case, I think we'd be better of decoding the wsgi.input data (into a str in Python 3 or into a unicode in Python 2) as soon as possible.
I'll look into it in the coming days (probably over the weekend).

**Comment by QuLogic** [comment]
The `wsgiref.handlers.read_environ()` suggestion refers to [this bit of code](https://github.com/geopython/pycsw/blob/master/pycsw/server.py#L59).

**Comment by ricardogsilva** [comment]
Hi @QuLogic, right, in that bit we are using `os.environ`, and that seems to be a no-no ;)

However, I'm thinking that that line is never actually used, because it seems that `pycsw.server.Csw()` is never called without the `env` variable, thus making that branch of the if statement moot.
Since `pycsw.server.Csw()` is only called from the `pycsw.wsgi.application()`, I'd say that the line that @QuLogic is pointing to is never actually used (and therefore we may want to remove it entirely).

I've also been looking at Python3's source code, specifically at the `wsgriref.handlers` implementation and it turns out that when using `wsgiref.handlers.CGIHandler`, as we are doing in `pycsw/csw.py` (our CGI wrapper), the `read_environ()` function is already being used internally.

To be continued...

**Comment by ricardogsilva** [comment]
@QuLogic, @tomkralidis it seems to me that the cause for the errors is in the test runner (`tests/run_tests.py`) and not in the main code:

* I set up a local pycsw instance at localhost:8000 using Python 2.7
* Made a POST request using `tests/suites/cite/post/19d2a6ed-be28-4866-ae15-e3bb634486cb.xml` (one of the offending files) as a payload
* The server replies correctly and I get none of those malformed XML errors
* Now do the same thing with Python 3.4
* The result is the same, no errors.

Now, go and run the test suite for python 3.4. The same errors that are seen in Travis appear. Strangely, if I run the same tests using Python 2.7 I get 232 tests failed (!), which does not seem to be the same as in Travis.

Looking at the test runner code in `tests/run_tests.py`, I've made some changes to the way that files are opened, mostly replacing the plain `open()` function with `codecs.open()` and using utf-8 as the encoding. I'm also encoding strings back to bytes using the same utf-8 encoding when they are sent to `urllib.request.Request` (and its Python2 counterpart). Now all tests run OK in Python 3.4. As I cannot alter your PR, please check my branch at 

https://github.com/ricardogsilva/pycsw/blob/pr419_py3k_order_fixes_fix/tests/run_tests.py

Strangely enough, when I run the tests for Python 2.7 I get the same 232 errors that appeared before my changes. When I look at the generated files in `tests/results` I can see that the errors are due to different namespace declarations. This seems like a totally unrelated issue. As I'm not too confortable with how the test suites are handled, I'll assume that @tomkralidis might have a clearer picture of what is going on here.

In conclusion, the changes present in my branch seem to fix the issues with malformed XML documents in both Python2.7 and 3.4. The test suite runs with zero errors for 3.4. As far as I can see, the errors reported for Python2.7 seem unrelated.

Your thoughts?

**Comment by tomkralidis** [comment]
@ricardogsilva thanks for the in depth analysis.  Indeed, updating the test runner fixes all the issues.  I believe the namespace declaration failures/issues are rooted in the underlying libxml2 library (I recall @kalxas and myself working through these to ensure something working on Travis).

I've cherry-picked your commit to the PR.  There were some unrelated errors that were due to a missing sort in the CSW3 code that are now locked down.

At this point #419 passes 100% in 2.6/2.7/3.4.  I'll test against CITE.

**Comment by tomkralidis** [comment]
Update: looks like we have some MIME type issues across versions when returning the `Content-Type` HTTP header:

Python 2: `application/xml; charset=UTF-8`
Python 3: `b'application/xml; charset=UTF-8`

Note that I caught this during CITE testing which skips the tests completely because of an invalid MIME type returned.

Any ideas/suggestions?

**Comment by tomkralidis** [comment]
I've put a workaround for this for the time being in https://github.com/geopython/pycsw/pull/419/commits/331ffa7ded0a079a322d511ddcf4b9a827934c34

@kalxas I propose to merge #419 given it works across CSW 2/3 versions for Python 2/3 in CITE.

**Comment by kalxas** [comment]
Thanks @tomkralidis @ricardogsilva @QuLogic 
+1 on merging #419
